### PR TITLE
Fixes #598

### LINKED ISSUE #598: Issue: `if` should not require `${{}}` for function calls / `!`
## Act version

```none
act version 9a7bc6
```

## Expected behaviour

```
      if: "contains(github.event_name, 'pull_request')"
```
should evaluate to false cleanly per https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsif
> When you use expressions in an `if` conditional, you may omit the expression syntax (`${{ }}`) because GitHub automatically evaluates the if conditional as an expression, unless the expression contains any operators. If the expression contains any operators, the expression must be contained within `${{ }}` to explicitly mark it for evaluation. For more information, see "[Context and expression syntax for GitHub Actions](https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions)."

and
```
      if: "!contains(github.event_name, 'pull_request')"
```
should evaluate to true

## Actual behaviour

it isn't evaluated as a function call w/o wrapping w/ `${{ ... }}`

```
[Spell checking/Spell checking]   ❌  Error in if: expression - checkout
DEBU[0001] expressions starting with ! must be wrapped in ${{ }} 
DEBU[0001] expressions starting with ! must be wrapped in ${{ }} 
Error: expressions starting with ! must be wrapped in ${{ }}
```

## Workflow and/or repository

<details>
  <summary>workflow</summary>

https://github.com/jsoref/spell-check-this/blob/4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03/.github/workflows/spelling.yml

(See also #597)
```none
name: Spell checking
on:
  pull_request_target:
  push:
  issue_comment:
  pull_request_review_comment:

jobs:
  build:
    name: Spell checking
    runs-on: ubuntu-latest
    steps:
    - name: checkout-merge
      if: "contains(github.event_name, 'pull_request')"
      uses: actions/checkout@v2.0.0
      with:
        ref: refs/pull/${{github.event.pull_request.number}}/merge
        fetch-depth: 5
    - name: checkout
      if: "!contains(github.event_name, 'pull_request')"
      uses: actions/checkout@v2.0.0
      with:
        fetch-depth: 5
    - uses: check-spelling/check-spelling@0.0.17-alpha
      with:
        experimental_apply_changes_via_bot: 1

```

</details>

## `act` output

<details>
  <summary>Log</summary>

```none
DEBU[0000] Loading environment from /Users/jsoref/code/spelling-org/spell-check-this/.env 
DEBU[0000] Loading secrets from /Users/jsoref/code/spelling-org/spell-check-this/.secrets 
DEBU[0000] Loading workflows from '/Users/jsoref/code/spelling-org/spell-check-this/.github/workflows' 
DEBU[0000] Reading workflow '/Users/jsoref/code/spelling-org/spell-check-this/.github/workflows/spelling.yml' 
DEBU[0000] Planning event: push                         
DEBU[0000] Loading slug from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0000] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0000] Loading revision from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0000] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0000] HEAD points to '4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03' 
DEBU[0000] HEAD matches refs/heads/prerelease           
DEBU[0000] using github ref: refs/heads/prerelease      
DEBU[0000] context env => map[ACT:true]                 
[Spell checking/Spell checking] 🚀  Start image=node:12.20.1-buster-slim
DEBU[0000] Loading slug from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0000] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0000] Loading revision from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0000] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0000] HEAD points to '4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03' 
DEBU[0000] HEAD matches refs/heads/prerelease           
DEBU[0000] using github ref: refs/heads/prerelease      
[Spell checking/Spell checking]   🐳  docker pull node:12.20.1-buster-slim
DEBU[0000] Image exists? true                           
[Spell checking/Spell checking] Removed container: 2535417245c6213ce3da11bfcfa92bcdfd16a7d3d3af0ec33646c02db074294a
[Spell checking/Spell checking]   🐳  docker volume rm act-Spell-checking-Spell-checking
[Spell checking/Spell checking]   🐳  docker create image=node:12.20.1-buster-slim platform=linux/amd64 entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[Spell checking/Spell checking] Created container name=act-Spell-checking-Spell-checking id=90b9071734c0adebfe3a95d79931a2791b1565adffbd6fac09673f2fe46c8bd8 from image node:12.20.1-buster-slim (platform: linux/amd64)
[Spell checking/Spell checking] ENV ==> [RUNNER_TOOL_CACHE=/opt/hostedtoolcache RUNNER_OS=Linux RUNNER_TEMP=/tmp]
[Spell checking/Spell checking]   🐳  docker run image=node:12.20.1-buster-slim platform=linux/amd64 entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[Spell checking/Spell checking] Starting container: 90b9071734c0adebfe3a95d79931a2791b1565adffbd6fac09673f2fe46c8bd8
[Spell checking/Spell checking] Started container: 90b9071734c0adebfe3a95d79931a2791b1565adffbd6fac09673f2fe46c8bd8
[Spell checking/Spell checking]   🐳  docker cp src=/Users/jsoref/code/spelling-org/spell-check-this/. dst=/Users/jsoref/code/spelling-org/spell-check-this
[Spell checking/Spell checking] Exec command '[mkdir -p /Users/jsoref/code/spelling-org/spell-check-this]'
DEBU[0000] Writing tarball /var/folders/r3/n29fz25x72x191fdv6mhhr3m0000gp/T/act049037753 from /Users/jsoref/code/spelling-org/spell-check-this/. 
DEBU[0000] Stripping prefix:/Users/jsoref/code/spelling-org/spell-check-this/ src:/Users/jsoref/code/spelling-org/spell-check-this/. 
[Spell checking/Spell checking] Extracting content from '/var/folders/r3/n29fz25x72x191fdv6mhhr3m0000gp/T/act049037753' to '/Users/jsoref/code/spelling-org/spell-check-this'
DEBU[0000] Writing entry to tarball workflow/event.json len:2 
DEBU[0000] Writing entry to tarball workflow/envs.txt len:0 
DEBU[0000] Writing entry to tarball home/.act len:0     
[Spell checking/Spell checking] Extracting content to '/Users/jsoref/code/spelling-org'
DEBU[0001] Loading slug from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0001] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0001] Loading revision from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0001] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0001] HEAD points to '4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03' 
DEBU[0001] HEAD matches refs/heads/prerelease           
DEBU[0001] using github ref: refs/heads/prerelease      
DEBU[0001] Loading slug from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0001] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0001] Loading revision from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0001] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0001] HEAD points to '4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03' 
DEBU[0001] HEAD matches refs/heads/prerelease           
DEBU[0001] using github ref: refs/heads/prerelease      
DEBU[0001] context env => map[ACT:true]                 
DEBU[0001] Loading slug from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0001] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0001] Loading revision from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0001] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0001] HEAD points to '4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03' 
DEBU[0001] HEAD matches refs/heads/prerelease           
DEBU[0001] using github ref: refs/heads/prerelease      
DEBU[0001] context env => map[ACT:true]                 
DEBU[0001] Evaluating 'github['event']['pull_request']['number']' instead of 'github.event.pull_request.number' 
ERRO[0001] Unable to interpolate string 'refs/pull/${{github.event.pull_request.number}}/merge' - [TypeError: Cannot access member 'number' of undefined] 
DEBU[0001] Loading slug from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0001] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0001] Loading revision from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0001] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0001] HEAD points to '4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03' 
DEBU[0001] HEAD matches refs/heads/prerelease           
DEBU[0001] using github ref: refs/heads/prerelease      
DEBU[0001] context env => map[ACT:true]                 
DEBU[0001] context env => map[ACT:true CI:true GITHUB_ACTION:0 GITHUB_ACTIONS:true GITHUB_ACTOR:nektos/act GITHUB_API_URL:https://api.github.com GITHUB_ENV:/Users/jsoref/code/spelling-org/workflow/envs.txt GITHUB_EVENT_NAME:push GITHUB_EVENT_PATH:/Users/jsoref/code/spelling-org/workflow/event.json GITHUB_GRAPHQL_URL:https://api.github.com/graphql GITHUB_REF:refs/heads/prerelease GITHUB_REPOSITORY:git@github.com:check-spelling/spell-check-this GITHUB_RUN_ID:1 GITHUB_RUN_NUMBER:1 GITHUB_SERVER_URL:https://github.com GITHUB_SHA:4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 GITHUB_TOKEN: GITHUB_WORKFLOW:Spell checking GITHUB_WORKSPACE:/Users/jsoref/code/spelling-org/spell-check-this INPUT_FETCH-DEPTH:5 INPUT_REF:refs/pull/${{github.event.pull_request.number}}/merge ImageOS:ubuntu20] 
DEBU[0001] Evaluating 'github['event']['pull_request']['number']' instead of 'github.event.pull_request.number' 
ERRO[0001] Unable to interpolate string 'refs/pull/${{github.event.pull_request.number}}/merge' - [TypeError: Cannot access member 'number' of undefined] 
[Spell checking/Spell checking] setupEnv => map[ACT:true CI:true GITHUB_ACTION:0 GITHUB_ACTIONS:true GITHUB_ACTOR:nektos/act GITHUB_API_URL:https://api.github.com GITHUB_ENV:/Users/jsoref/code/spelling-org/workflow/envs.txt GITHUB_EVENT_NAME:push GITHUB_EVENT_PATH:/Users/jsoref/code/spelling-org/workflow/event.json GITHUB_GRAPHQL_URL:https://api.github.com/graphql GITHUB_REF:refs/heads/prerelease GITHUB_REPOSITORY:git@github.com:check-spelling/spell-check-this GITHUB_RUN_ID:1 GITHUB_RUN_NUMBER:1 GITHUB_SERVER_URL:https://github.com GITHUB_SHA:4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 GITHUB_TOKEN: GITHUB_WORKFLOW:Spell checking GITHUB_WORKSPACE:/Users/jsoref/code/spelling-org/spell-check-this INPUT_FETCH-DEPTH:5 INPUT_REF:refs/pull//merge ImageOS:ubuntu20]
DEBU[0001] Evaluating 'Boolean(contains(github['event_name'], 'pull_request'))' instead of 'Boolean(contains(github.event_name, 'pull_request'))' 
DEBU[0001] expression 'contains(github.event_name, 'pull_request')' evaluated to 'false' 
DEBU[0001] Skipping step 'checkout-merge' due to 'contains(github.event_name, 'pull_request')' 
DEBU[0001] Loading slug from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0001] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0001] Loading revision from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0001] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0001] HEAD points to '4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03' 
DEBU[0001] HEAD matches refs/heads/prerelease           
DEBU[0001] using github ref: refs/heads/prerelease      
DEBU[0001] Loading slug from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0001] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0001] Loading revision from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0001] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0001] HEAD points to '4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03' 
DEBU[0001] HEAD matches refs/heads/prerelease           
DEBU[0001] using github ref: refs/heads/prerelease      
DEBU[0001] context env => map[ACT:true]                 
DEBU[0001] Loading slug from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0001] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0001] Loading revision from git directory '/Users/jsoref/code/spelling-org/spell-check-this/.git' 
DEBU[0001] Found revision: 4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 
DEBU[0001] HEAD points to '4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03' 
DEBU[0001] HEAD matches refs/heads/prerelease           
DEBU[0001] using github ref: refs/heads/prerelease      
DEBU[0001] context env => map[ACT:true]                 
DEBU[0001] context env => map[ACT:true CI:true GITHUB_ACTION:1 GITHUB_ACTIONS:true GITHUB_ACTOR:nektos/act GITHUB_API_URL:https://api.github.com GITHUB_ENV:/Users/jsoref/code/spelling-org/workflow/envs.txt GITHUB_EVENT_NAME:push GITHUB_EVENT_PATH:/Users/jsoref/code/spelling-org/workflow/event.json GITHUB_GRAPHQL_URL:https://api.github.com/graphql GITHUB_REF:refs/heads/prerelease GITHUB_REPOSITORY:git@github.com:check-spelling/spell-check-this GITHUB_RUN_ID:1 GITHUB_RUN_NUMBER:1 GITHUB_SERVER_URL:https://github.com GITHUB_SHA:4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 GITHUB_TOKEN: GITHUB_WORKFLOW:Spell checking GITHUB_WORKSPACE:/Users/jsoref/code/spelling-org/spell-check-this INPUT_FETCH-DEPTH:5 ImageOS:ubuntu20] 
[Spell checking/Spell checking] setupEnv => map[ACT:true CI:true GITHUB_ACTION:1 GITHUB_ACTIONS:true GITHUB_ACTOR:nektos/act GITHUB_API_URL:https://api.github.com GITHUB_ENV:/Users/jsoref/code/spelling-org/workflow/envs.txt GITHUB_EVENT_NAME:push GITHUB_EVENT_PATH:/Users/jsoref/code/spelling-org/workflow/event.json GITHUB_GRAPHQL_URL:https://api.github.com/graphql GITHUB_REF:refs/heads/prerelease GITHUB_REPOSITORY:git@github.com:check-spelling/spell-check-this GITHUB_RUN_ID:1 GITHUB_RUN_NUMBER:1 GITHUB_SERVER_URL:https://github.com GITHUB_SHA:4101393f8d4d21cbaadaf88c3ac1b7ad3754bc03 GITHUB_TOKEN: GITHUB_WORKFLOW:Spell checking GITHUB_WORKSPACE:/Users/jsoref/code/spelling-org/spell-check-this INPUT_FETCH-DEPTH:5 ImageOS:ubuntu20]
[Spell checking/Spell checking]   ❌  Error in if: expression - checkout
DEBU[0001] expressions starting with ! must be wrapped in ${{ }} 
DEBU[0001] expressions starting with ! must be wrapped in ${{ }} 
Error: expressions starting with ! must be wrapped in ${{ }}
```

</details>

### PR DESCRIPTION
**Pull request by Renstrom**
# Issue solution for issue [#598](https://github.com/nektos/act/issues/598)
fixes #598 
## Problem description
This solution fixes issue #598, the problem occurs when the workflow got initiated 
model/workflow.go:325 (func ReadWorkFlow()). The reason was due to that the decoder removes the ("") when it got set into the interface. 
## Solution
This solution ${[ }} to the corners if "" statements exist.
So if the problem is:
"!contains(github.event_name, 'pull_request')" 
then the string becomes:
 ${{ !contains(github.event_name, 'pull_request' }}. 
This makes it so the decoder interprets the string to ${{!contains(github.event_name, 'pull_request')}} which solves the issue.

## Alternative solution
I tried looking for a way to optimize the code by fixing the yaml decoder, but couldn't find a better solution than this.

### CODE REVIEW

**Comment by catthehacker** [comment]
I'm not sure about the whole "tmp file" thing, can't it be done in memory?

**Comment by Renstrom** [comment]
So far I haven't found a solution for this based on the structure of act. Since I believe the problem lies in the YML decoder way of preserving quotes, (I'm not that familiar with it). Since in the if statement "", '' and non-quotes gets interpreted as a normal string.  Which makes:

"Hello World" => "Hello World"

'Hello World' => "Hello World"

Hello World => "Hello World"

I can try to look more into detail on how to preserve the quotes without using memory tomorrow and see if I can find any solutions to it.

**Comment by Renstrom** [comment]
I changed the solution, to not create a temporary file, instead comparing lines to the yml file and using regex to replace the values after assignment. 

I tried finding a more optimized way with yaml decoder, but to no awail.

**Comment by catthehacker** [comment]
Sorry, didn't had time to review the PR yet but will definitely check it out. Thanks for contributing.

**Comment by Renstrom** [comment]
Found an issue with the solution while testing it. From what I gather is that in If statements ${{boolean}} behaves the same as "" so i decided to change it to this. Since there were some evaluation problems with evaluating "\ \" making the statements always true.

**Comment by Renstrom** [comment]
@catthehacker Do you want me to add tests for amd/linux64 aswell ?

**Comment by catthehacker** [comment]
I've reworked tests few days ago so no changes are needed 😺, just need to PR them.

**Comment by Renstrom** [comment]
@catthehacker Found an optimization to the solution. This is the finalized version of my solution. If there is are any questions feel free to ask.

**Comment by cplee** [comment]
@Renstrom  - need to resolve a conflict, and this one is ready to go

**Comment by Renstrom** [comment]
> @Renstrom - need to resolve a conflict, and this one is ready to go

Will do, I can probably fix it later today.

**Comment by Renstrom** [comment]
> 
> 
> @Renstrom - need help resolving a conflict

Due to the merge conflict that I fixed, I managed to get some errors. I'll look into it now in order to fix it with the current build

File: pkg/model/planner.go [hunk 1]
```diff
@@ -121,20 +157,41 @@ func NewWorkflowPlanner(path string, noWorkflowRecurse bool) (WorkflowPlanner, e
...
 		ext := filepath.Ext(wf.workflowFileInfo.Name())
 		if ext == ".yml" || ext == ".yaml" {
+
 			f, err := os.Open(filepath.Join(wf.dirPath, wf.workflowFileInfo.Name()))
 			if err != nil {
 				return nil, err
 			}
 
-			log.Debugf("Reading workflow '%s'", f.Name())
+			log.Debugf("Reading workflow '%s'", file.Name())
```
Comment (catthehacker) [inline pkg/model/planner.go:166]: Shouldn't it be `f.Name()`?

File: pkg/runner/runner_test.go [hunk 2]
```diff
@@ -96,11 +96,12 @@ func TestRunEvent(t *testing.T) {
 		{"testdata", "workdir", "push", "", platforms, ""},
 		{"testdata", "defaults-run", "push", "", platforms, ""},
 		{"testdata", "uses-composite", "push", "", platforms, ""},
+    {"testdata", "issue-598", "push", "", platforms, ""},
```
Comment (catthehacker) [inline pkg/runner/runner_test.go:99]: small formatting issue

**Review by catthehacker — CHANGES_REQUESTED** [review]

File: pkg/model/planner.go (hunk 1 above)
Comment (Renstrom) [reply pkg/model/planner.go:166]: Yeah, didn't notice the change in the syntax of debug file names. Ill resolve this

File: pkg/model/planner.go [hunk 3]
```diff
@@ -135,6 +169,18 @@ func NewWorkflowPlanner(path string, noWorkflowRecurse bool) (WorkflowPlanner, e
...
+				f.Close()
+				return nil, errors.WithMessagef(err, "error occuring when resetting io pointer, %s", wf.workflowFileInfo.Name())
+			}
+			log.Debugf("Correcting if statements '%s'", f.Name())
+			content, err := ioutil.ReadFile(filepath.Join(wf.dirPath, wf.workflowFileInfo.Name()))
+			if err != nil {
+				return nil, errors.WithMessagef(err, "error occuring when reading file, %s", wf.workflowFileInfo.Name())
+			}
+
+			err = FixIfStatement(content, workflow)
```
Comment (catthehacker) [inline pkg/model/planner.go:183]:
```suggestion
			err = FixIfStatement(content, workflow)
			if err != nil {
				return err
			}
```

**Review by catthehacker — APPROVED** [review]

File: pkg/model/planner.go [hunk 4]
```diff
@@ -50,6 +50,40 @@ func (r *Run) Job() *Job {
 	return r.Workflow.GetJob(r.JobID)
 }
 
+// Helper function for FixIfstatement
+func FixIfStatement1(val string, lines [][][]byte, l int) (string, error) {
+	if val != "" {
+		line := lines[l-1][0]
+		outcome := regexp.MustCompile(`\s+if:\s+".*".*`).FindSubmatch(line)
```
Comment (jsoref) [inline pkg/model/planner.go:57]: is:
```
if:"foo"
```

not valid?

File: pkg/runner/testdata/issue-598/spelling.yml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        uses: actions/hello-world-javascript-action@main
+        with:
+          who-to-greet: 'Act the Octocat'
+      - name: My third false step
+        if: "endsWith('Hello world', 'o2')"
+        uses: actions/hello-world-javascript-action@main
+        with:
+          who-to-greet: 'Git the Octocat'
+           
+
```
Comment (jsoref) [inline pkg/runner/testdata/issue-598/spelling.yml:31]: The strange combination of blankish lines and no newline at eof is strange

File: pkg/model/planner.go (hunk 4 above)
Comment (Renstrom) [reply pkg/model/planner.go:57]: While I did my bugtesting, I both checked the documentation and tried each if statement combination that I could think of. The if:"foo" was in the incorrect format according to github action when i tested it. Which is why I demand the \s+ before the "foo"

**Pull request merged**